Fix punished scripts stopping behavior and application shutdown while scripts being punished

ZappoMan opened the pull request.

ZappoMan commented:
Test Plan:
- Smoke test - make sure performance hasn't regressed
- Run a really bad script:
```
// an intentionally bad script to test the punish behavior

Script.update.connect(function(deltaTime){
    print("time since last update:" + deltaTime);
    var a = 1234567;
    var b = 89012345;
    for (var i = 0; i < 100000000; i++) {
        var c = a * b;
    }
});
```
- Make sure that while that really bad script is running, that you can still:
    - Close the app with Ctrl+Q and the red X button
    - Stop the really bad script from the running scripts dialog box

Notice with the current Dev-builds -- that exiting the program and closing the scripts will take a very long time.

hifi-gustavo commented:
Build console output for commit: 755989b8ffe1d3afe8683d552f5f89d47ca973d5 available at: https://gist.github.com/64b5b51b56039f2a1a4b7a20a8ca9cbe<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8328-755989b.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8328-755989b.dmg

OmegaHeron commented:
Testing Windows 10 - i5 6500 CPU - GTX970 GPU

Using DEV-5211 running bad script I can't say that I see either a huge delay in script termination or closure of Interface.  But - if I have default script set + bad script running and remove all scripts - bad script certainly takes longer to stop (1 second or so).  Same with closing with Ctrl Q or X.

Using this PR running bad script - behaves sames as ^^.  It looks like someone with hardware config that can repro issue below will need to test.  

"Notice with the current Dev-builds -- that exiting the program and closing the scripts will take a very long time."

I see no regression to performance original PR addressed - this PR continues to offer improved experience on my i5.

howard-stearns commented:
Running on my MacBook Pro (see ticket), this works as far as stopping an individual script from the running scripts dialog. However, the Remove All button doesn't work. 

repro: run this pr on an MBP, with default scripts. Wait for load. (Get about 10-11 fps.) Open Running Scripts and press Remove All button button. Nothing happens, even after several minutes. 

Clicking an individual script's 'X' button removes the script within a second. You can even run just one script -- say, dialTone.js -- and kill it individually. But running that same single script and using the Remove All button has no effect.

ZappoMan commented:
@howard-stearns - hmm - ok - I have another idea...

ZappoMan commented:
test this please

zzmp commented on libraries/script-engine/src/ScriptEngine.cpp:
```diff
@@ -845,14 +845,29 @@ void ScriptEngine::run() {
         // time into the future.
-        auto wouldSleep = (sleepUntil - clock::now());
+        auto wouldSleep = (sleepUntil - beforeSleep);
         auto avgerageUpdate = totalUpdates / thisFrame;
```
s/avgerage/average

ZappoMan replied:
fixing

zzmp commented on libraries/script-engine/src/ScriptEngine.cpp:
```diff
@@ -845,14 +845,29 @@ void ScriptEngine::run() {
+            if (wouldSleepSlice > FRAME_DURATION) {
+                thisSleepUntil = clock::now() + FRAME_DURATION;
+            }
+            std::this_thread::sleep_until(thisSleepUntil);
```
You are sleeping for the `std::min(sleepUntil, clock::now() + FRAME_DURATION)` to avoid blocking for more than a frame at a time?

Can you replace this logic with `std::min(...)`? It seems easier to parse.

ZappoMan replied:
I will try - I'm not sure that chrono's support min()

zzmp replied:
std::chrono::time_point implements operator< so it should just work.

zzmp commented on libraries/script-engine/src/ScriptEngine.cpp:
```diff
@@ -845,14 +845,29 @@ void ScriptEngine::run() {
+                thisSleepUntil = clock::now() + FRAME_DURATION;
+            }
+            std::this_thread::sleep_until(thisSleepUntil);
+            firstTime = false;
```
This is related to the logic in L860-862 more than it is related to the sleep logic. Can you move it up to L863 to make it easier to follow?

zzmp commented:
Why do you not process events the first time? Why did you put process events before the sleep?

ZappoMan commented:
@zzmp - later in the regular processing we do process events...

ZappoMan commented:
@birarda @howard-stearns @zzmp - I believe I've addressed all your feedback:
- The logic should be much more clear (using `std::min()` and `std::max()`)
- The processEvents should allow shutdown even while in a "punish" script case
- The logic should now work even if there is no entity server and therefore physics will never get enabled

hifi-gustavo commented:
Build console output for commit: e879f22c952772658276153fb3bd8690affe31fe available at: https://gist.github.com/64b5b51b56039f2a1a4b7a20a8ca9cbe<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8328-e879f22.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8328-e879f22.dmg

sethalves commented:
I haven't done a lot of testing here, but I can say that the ambient-sound AC-script that eric wrote for playa (ACAudioSearchAndInject.js) fails in master and works with this PR.

OmegaHeron commented:
Windows 10 - i5 6500 CPU - GTX970 GPU

General smoke test = all pass.
Criteria for bad script test = all pass.
In ref to no entity server present.  As far as I can tell, yes - signals are passed.
